### PR TITLE
adds wait with exponential backoff to create ObjectUser

### LINKED ISSUE #2490: Ceph Object Store User's Secret not created if applied too soon after Ceph Object Store
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
If a manifest for the CephObjectStoreUser is created right after applying the manifest CephObjectStore and the CephObjectStoreUser had not been created by the operator yet, then the CephObjectStoreUser secret will not be created at all.

Applying the manifest again after the CephObjectStore is up will create the secret either. Deleting the user and applying it again will create the secret. 

**Expected behavior:**
Expect behavior is for the rook operator to handle the sequence and queue the CephObjectStoreUser  until its dependency on CephObjectStore has been met. 

**How to reproduce it (minimal and precise):**
On a healthy rook-ceps system without an existing CephObjectStore or CephObjectStoreUser created:

1. create the manifests for the ObjectStore and User as per https://rook.io/docs/rook/v0.9/ceph-object.html ( or use the example manifest for the release )
 
2. apply the manifests for the ObjectStore as per https://rook.io/docs/rook/v0.9/ceph-object.html 
-  kubectl apply -f https://raw.githubusercontent.com/rook/rook/master/cluster/examples/kubernetes/ceph/object.yaml

3. before the rook operator has created the ObjectStore, apply the manifests for the ObjectStoreUser as per https://rook.io/docs/rook/v0.9/ceph-object.html 
- kubectl apply -f https://raw.githubusercontent.com/rook/rook/master/cluster/examples/kubernetes/ceph/object-user.yaml

 
**Environment**:
* OS (e.g. from /etc/os-release): not os specific
* Kernel (e.g. `uname -a`): not kernel specific
* Cloud provider or hardware configuration: not cloud specific 
* Rook version v0.9.0
* Kubernetes version 1.12
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): not cluster specific 
* Storage backend status: Ceph is healthy, not health specific

### PR DESCRIPTION
**Pull request by umangachapagain**
Signed-off-by: Umanga Chapagain <chapagainumanga@gmail.com>

**Description of your changes:**
creating an objectstore user requires that the objectstore is
created first. This commits adds changes to exponentially wait
for the object store to be created before the objectstore user
can be created.

**Which issue is resolved by this Pull Request:**
Resolves #2490

**Checklist:**
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../CONTRIBUTING.md#comments)

### CODE REVIEW

File: pkg/operator/ceph/object/user/controller.go [hunk 1]
```diff
@@ -152,6 +173,21 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+	err := wait.ExponentialBackoff(backoff, func() (ok bool, err error) {
+		_, err = context.RookClientset.CephV1().CephObjectStores(u.Namespace).Get(u.Spec.Store, metav1.GetOptions{})
+		if err == nil {
+			return true, nil
+		}
+		// wait only if objectStore is not found. return on any other error
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return true, err
```
Comment (leseb) [inline pkg/operator/ceph/object/user/controller.go]:
```suggestion
		return true, nil
```

File: pkg/operator/ceph/object/user/controller.go [hunk 2]
```diff
@@ -152,6 +173,21 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
 	}
 	objContext := object.NewContext(context, u.Spec.Store, u.Namespace)
 
+	// Wait with exponential backoff for the objectstore to be created
+	err := wait.ExponentialBackoff(backoff, func() (ok bool, err error) {
+		_, err = context.RookClientset.CephV1().CephObjectStores(u.Namespace).Get(u.Spec.Store, metav1.GetOptions{})
+		if err == nil {
```
Comment (leseb) [inline pkg/operator/ceph/object/user/controller.go]: why not removing this condition and let the last return returning `true, nil`?

File: pkg/operator/ceph/object/user/controller.go [hunk 3]
```diff
@@ -152,6 +173,21 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+			return true, nil
+		}
+		// wait only if objectStore is not found. return on any other error
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return true, err
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get objectstore %s. error: %+v", u.Spec.Store, err)
```
Comment (leseb) [inline pkg/operator/ceph/object/user/controller.go]:
```suggestion
		return fmt.Errorf("failed to get objectstore %s. %+v", u.Spec.Store, err)
```

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: I was thinking of 3 conditions,
1. When there is no error (return true and stop waiting immediately)
2. When there is an IsNotFound error (return false and wait)
3. When there is any other error (return true with error to exit out of wait and let admin handle the issue because this might not something that we expect to self heal)

Makes sense?

File: pkg/operator/ceph/object/user/controller.go [hunk 4]
```diff
@@ -152,6 +173,21 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
 	}
 	objContext := object.NewContext(context, u.Spec.Store, u.Namespace)
 
+	// Wait with exponential backoff for the objectstore to be created
+	err := wait.ExponentialBackoff(backoff, func() (ok bool, err error) {
+		_, err = context.RookClientset.CephV1().CephObjectStores(u.Namespace).Get(u.Spec.Store, metav1.GetOptions{})
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go]: This will only wait for the object store CR to exist. In addition we will need to check that rook has initialized the object store. When rook is initializing the object store the last thing it will do is [start the rgw pod(s)](https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/rgw.go#L95), so you can check that the rgw pod is in the Running state. You could query for pods matching the label `rgw=<store name>` and `app=rook-ceph-rgw`.

File: pkg/operator/ceph/object/user/controller.go [hunk 5]
```diff
@@ -152,6 +173,21 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+		if err == nil {
+			return true, nil
+		}
+		// wait only if objectStore is not found. return on any other error
+		if errors.IsNotFound(err) {
+			return false, nil
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go]: It would be good to add logging so we can see that the operator is waiting for the object store to exist

File: pkg/operator/ceph/object/user/controller.go (hunk 4 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: can we make use of https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/rgw.go#L303 instead?
To check for rgw pods, I'd have to know the name which might change. Solution would be to Query for deployments. This is what the above mentioned function does.

File: pkg/operator/ceph/object/user/controller.go (hunk 5 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: ack

File: pkg/operator/ceph/object/user/controller.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go]: That checks if it is configured, but not if it is running. If you check for a running rgw pod then it will give you more accurate info about the service being available. You don't actually need to know the name of the pod. If you call `List()` on pods you can give it the `ListOptions`, which allows you to match based on the labels.

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go]: The conditions make sense, i think @leseb is wondering if there is a simpler way to write those conditions. This seems like the simplest, although functionally the same as what you have:
```
if err != nil && errors.IsNotFound(err) {
  return false, nil
}
// err might be nil or non-nil
return true, err
```

Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: ack.

File: pkg/operator/ceph/object/user/controller.go (hunk 4 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: ack.

File: pkg/operator/ceph/object/user/controller.go [hunk 6]
```diff
@@ -151,7 +171,37 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return true, fmt.Errorf("CephObjectStore %s may not be initialized. %+v", u.Spec.Store, err)
+		}
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed while waiting for objectstore %s to be ready. %+v", u.Spec.Store, err)
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go]: How about if we just log the error, and allow the method to proceed and attempt to create the user? This wait method isn't technically critical for the user to be created.

File: pkg/operator/ceph/object/user/controller.go [hunk 7]
```diff
@@ -151,7 +171,37 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+			return true, fmt.Errorf("CephObjectStore %s could not be found. %+v", u.Spec.Store, err)
+		}
+		// check if ObjectStore is initialized
+		// rook does this by starting the RGW pod(s)
+		selector := fmt.Sprintf("%s=%s,%s=%s",
+			"rgw", u.Spec.Store,
+			"app", "rook-ceph-rgw",
+		)
+		logger.Infof("waiting for CephObjectStore %s to be initialized", u.Spec.Store)
+		_, err = context.Clientset.CoreV1().Pods(u.Namespace).List(metav1.ListOptions{LabelSelector: selector})
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go]: Could you also look at the pods and verify that at least one is in the `Running` state? It typically takes a few extra seconds for the pod to start.

File: pkg/operator/ceph/object/user/controller.go (hunk 6 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: I thought it was. Because user creation depends on creation of object store?
If we just wait for sometime and proceed anyway and the creation still fails, isn't that bad?

Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go]: User creation will likely fail if this fails, but at least the operator can attempt the creation before giving up.

Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: ack.

File: pkg/operator/ceph/object/user/controller.go [hunk 8]
```diff
@@ -151,7 +171,46 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
 		DisplayName: &displayName,
 	}
 	objContext := object.NewContext(context, u.Spec.Store, u.Namespace)
-
+	// Wait with exponential backoff for the objectstore to be created
+	err := wait.ExponentialBackoff(backoff, func() (ok bool, err error) {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: Please start the `func()` on a new line for readability.

File: pkg/operator/ceph/object/user/controller.go [hunk 9]
```diff
@@ -40,6 +44,22 @@ const (
...
 
+const (
+	waitDuration = time.Minute * 1
+	waitFactor   = 2
+	waitJitter   = 0.5
+	waitSteps    = 4
+	waitCap      = time.Minute * 15
+)
+
+var backoff = wait.Backoff{
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: I don't see a need to define these as unexported constants and only use them in this one place. I would rather see `var backoff = wait.Backoff{` set directly. Additionally, it would be good to note  what the effect of the options selected is intended to be for those not familiar with all the Backoff parameters.

File: pkg/operator/ceph/object/user/controller.go [hunk 10]
```diff
@@ -151,7 +171,46 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			logger.Errorf("CephObjectStore %s may not be initialized. %+v", u.Spec.Store, err)
+			return true, err
+		}
+		// check if atleast one pod is running
+		for _, pod := range pods.Items {
+			if pod.Status.Phase == v1.PodRunning {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: Why not use the `FieldSelector` in `ListOptions` to only list running pods instead of implementing this check manually?

File: pkg/operator/ceph/object/user/controller.go [hunk 11]
```diff
@@ -151,7 +171,46 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+				logger.Infof("wait condition for CephObjectStore %s met", u.Spec.Store)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: I'm not sure users will understand what this means. I'd suggest "CephObjectStore %s has been created successfully"

File: pkg/operator/ceph/object/user/controller.go [hunk 12]
```diff
@@ -151,7 +171,46 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+			logger.Errorf("CephObjectStore %s could not be found. %+v", u.Spec.Store, err)
+			return true, err
+		}
+		// check if ObjectStore is initialized
+		// rook does this by starting the RGW pod(s)
+		selector := fmt.Sprintf("%s=%s,%s=%s",
+			"rgw", u.Spec.Store,
+			"app", "rook-ceph-rgw",
+		)
+		logger.Infof("waiting for CephObjectStore %s to be initialized", u.Spec.Store)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: Should this be an info message or debug? I would lean towards debug.

File: pkg/operator/ceph/object/user/controller.go [hunk 13]
```diff
@@ -151,7 +171,46 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+		for _, pod := range pods.Items {
+			if pod.Status.Phase == v1.PodRunning {
+				logger.Infof("wait condition for CephObjectStore %s met", u.Spec.Store)
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		logger.Errorf("failed while waiting for objectstore %s to be ready. %+v", u.Spec.Store, err)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: s/failed/timed out/

File: pkg/operator/ceph/object/user/controller.go [hunk 14]
```diff
@@ -151,7 +171,46 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+				return false, nil
+			}
+			logger.Errorf("CephObjectStore %s could not be found. %+v", u.Spec.Store, err)
+			return true, err
+		}
+		// check if ObjectStore is initialized
+		// rook does this by starting the RGW pod(s)
+		selector := fmt.Sprintf("%s=%s,%s=%s",
+			"rgw", u.Spec.Store,
+			"app", "rook-ceph-rgw",
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: Use `appName` here instead of hardcoding `rook-ceph-rgw`. Use k8sutil.AppAttr here instead of hardcoding `app`.

File: pkg/operator/ceph/object/user/controller.go (hunk 12 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: Info is better. Provides info to the admin about what is happening. If we keep it as debug, admin might assume the `wait` as a `crash` or `stuck` deployment unless he switches to debug (which is not default).

File: pkg/operator/ceph/object/user/controller.go (hunk 9 above)
Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: Even my understanding is not so clear (to document better). Any suggestions would be appreciated.

Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go]: - Sounds good to define this `backoff` var inside the method instead of as a package variable
- It would be nice if there were several retries within the first minute as well since it's likely to succeed in that timeframe if the object store CR is created at the same time. Perhaps just changing the duration will be sufficient: `Duration: time.Second * 15`
- You can find the docs here on the backoff params. If you could just add a comment about what these mean it would be helpful. https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Backoff

Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: Refined the params and added comments.

File: pkg/operator/ceph/object/user/controller.go [hunk 15]
```diff
@@ -151,7 +155,62 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+	//
+	// It checks the condition up to `Steps` times, increasing the wait by multiplying
+	// the previous `Duration` by `Factor`
+	//
+	// If `Jitter` is greater than zero, a random amount of each duration is added
+	// Jitter is applied after `Cap`
+	//
+	// If the condition never returns true, ErrWaitTimeout is returned. All errors
+	// except IsNotFound terminates immediately
+	err := wait.ExponentialBackoff(wait.Backoff{
```
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go]: - I thought Steps is the number of retries it will increase the backoff period. So if we had `Steps: 3`, the settings would currently retry after 15, 30, 60, 60, 60, ... until it hits the Cap of 5 minutes. 
- For the comment, could you summarize the behavior? For example, if we just say something like "Will retry several times using exponential backoff, starting after 15 seconds and retrying several times until the timeout of 5 minutes is reached."

Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: + Based on your assumption, it would no longer be an exponential backoff at one point.
+ Summaried with Example on it's behavior. This should be enable to backtrace later on, if required.

Comment (BlaineEXE) [reply pkg/operator/ceph/object/user/controller.go]: The documentation for `struct wait.Backoff` does say something different than the documentation for `func wait.ExponenetialBackoff`. We should document this here since it has confused both Travis and myself.

File: pkg/operator/ceph/object/user/controller.go [hunk 16]
```diff
@@ -151,7 +155,63 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
...
+	// If the condition never returns true, ErrWaitTimeout is returned
+	// All errors except IsNotFound terminates immediately
+	//
+	// On Step 1 we'll wait for 15 sec (i.e. Duration)
+	// On Step 2 we'll wait for 15*1.5=22.5 sec (i.e. Previous Duration * Factor)
+	// On Step 3 we'll wait for 15*(1.5^2)=33.75 sec (i.e. Duration * (Factor^(Current Step-1)) )
+	// Once we have this duration > Cap (i.e 5 mins) or Step=10, we stop waiting
+	//
+	// In this configuration, we will wait for 5 mins making 8 retries with exponential backoff
+	err := wait.ExponentialBackoff(wait.Backoff{
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go]: I think this exponential backoff is potentially dangerous for Rook. We are wanting to wait some number of minutes for a case to be true, but with the exponential backoff, this seems to be implemented under the hood with a dumb timeout, and the condition is not checked one last time after the timeout. In the example I made to test this more quickly, I want to wait 10 seconds for a condition, but it only actually tests it up to 5.0625 seconds. This seems to me to defeat what we are trying to achieve in Rook.

And based on my math, this configuration will only test up to 256.2891 seconds, which is over 30 seconds shorter than the desired 5 minute timeout. And because of the exponential equation, this becomes difficult to reason about. 

Why don't we either (A) use a linear timeout that we can reason about more easily or (B) use a watcher that will put less pressure on the API server?

Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go]: It was my suggestion to use the exponential backoff so you can blame me :). In many places K8s is using a backoff scheme to ensure the api server is not being overwhelmed. That said, it does seem to be more complexity than we need and causes all these questions around the behavior. 

If we're using a relatively large polling interval, it would certainly be simpler to follow what we're doing other places and just query with the same interval with `wait.Poll()` and keep it simple. As long as it's only polling from a single place (the operator) rather than all pods, it's not going to overwhelm the API server.  What if we just poll every 15 seconds for 5 minutes? Nice and simple. 

@umangachapagain What do you think? Thanks for your patience while we talk through whether a new pattern is the right thing...

Comment (umangachapagain) [reply pkg/operator/ceph/object/user/controller.go]: > implemented under the hood with a dumb timeout 

Can you explain this? The way I see it, we have a set of parameters that we can configure (including the dumb timeout). Based on these parameters, a predefined mathematical function is used to create a "Wait".

> and the condition is not checked one last time after the timeout

I believe this is literally what a "timeout" is supposed to do.

> this configuration will only test up to 256.2891 seconds, which is over 30 seconds shorter than the desired 5 minute timeout

we could easily add "30 sec" to Cap and get back the lost seconds. If that seems problematic.

> use a linear timeout that we can reason about more easily

Linear or Exponential both are reasonable mathematical functions. And if my understanding is correct Exponential Timeout was designed to better handle the server load without always making the same assumptions (Polling every 10 or 15 second even after knowing that Pod could take more than a minute to come up, wasting at least 4 retries).

> What if we just poll every 15 seconds for 5 minutes?

If this is desired, setting the Factor to 1 will give you the same effect.

Here is my reasoning for why Exponential Backoff is good, based on the configuration I have used,
1. First wait (15 sec) is ideal to wait for the ObjectStore CR to be created after ObjectUser Cr is created. If for any reason this does not happen, Second Wait (22.5 sec) for Objectstore CR to appear making it a total wait of 37.5 sec for both CRs to be applied sounds reasonable to me.
2. After this happens, third & fourth wait (totaling >1 min) should be enough for a rgw pod to be running, unless there is a major performance issue.
3. So, by 4th wait we should be good and exiting the wait loop. If by any chance that doesn't happen, 5th 6th 7th 8th waits (each wait between 1.2 mins - 4.2mins) should be more than sufficient for all the required resources to come up and we should be good to proceed without having to worry about the 30sec lost.

That being said, if you still think linear wait or watch is better, I will happily abandon this work. This is a good discussion and so much to learn :)

Comment (travisn) [reply pkg/operator/ceph/object/user/controller.go]: The exponential backoff will certainly work, there is no doubt. Here are the downsides to it that I'm seeing:
- Developer complexity/code readability is more difficult. It's just harder to understand.
- There isn't really a need for it. The need for exponential backoff is when there are many clients that would be flooding the server. In this case, it's just a single client and the api server will not be overloaded by this request.
- As the backoff wait time increases with each try, you have the potential for waiting longer unnecessarily to create the resource after it's ready to be created. A linear approach of 15s will likely create the resource sooner.

**Comment by umangachapagain** [comment]
> @umangachapagain Thoughts on this? I'd like to get it into v1.1. Thanks!
>
Totally forgot about this. Will put a fix ASAP.
